internal/pkg/csumio/csumreader_test.go

```diff
@@ -184,3 +184,234 @@ func TestReader_CorrectChecksum_VerificationEnabled(t *testing.T) {
+				totalLen = chunkSize + readLen - 1
+				preConsume = chunkSize
+			} else {
+				totalLen = preConsume + readLen - 1
+			}
+		}
+
+		return readLen, preConsume, totalLen
+	}
+
+	cases := []matrixCase{
+		{
+			name:               "case01_pGreater_remainingGreater_innerBufferEmpty",
+			bufferLen:          bufferLenGreater,
+			remainingData:      remainingGreater,
+			isInnerBufferEmpty: true,
+		},
+		{
+			name:               "case02_pGreater_remainingEqual_innerBufferPartial",
+			bufferLen:          bufferLenGreater,
+			remainingData:      remainingEqual,
+			isInnerBufferEmpty: false,
+		},
+		{
+			name:               "case03_pGreater_remainingLess_innerBufferEmpty",
+			bufferLen:          bufferLenGreater,
+			remainingData:      remainingLess,
+			isInnerBufferEmpty: true,
+		},
+		{
+			name:               "case04_pEqual_remainingGreater_innerBufferPartial",
+			bufferLen:          bufferLenEqual,
+			remainingData:      remainingGreater,
+			isInnerBufferEmpty: false,
+		},
+		{
+			name:               "case05_pEqual_remainingEqual_innerBufferEmpty",
+			bufferLen:          bufferLenEqual,
+			remainingData:      remainingEqual,
+			isInnerBufferEmpty: true,
+		},
+		{
+			name:               "case06_pEqual_remainingLess_innerBufferPartial",
+			bufferLen:          bufferLenEqual,
+			remainingData:      remainingLess,
+			isInnerBufferEmpty: false,
+		},
+		{
+			name:               "case07_pSmaller_remainingGreater_innerBufferEmpty",
+			bufferLen:          bufferLenSmaller,
+			remainingData:      remainingGreater,
+			isInnerBufferEmpty: true,
+		},
+		{
+			name:               "case08_pSmaller_remainingEqual_innerBufferPartial",
+			bufferLen:          bufferLenSmaller,
+			remainingData:      remainingEqual,
+			isInnerBufferEmpty: false,
+		},
+		{
+			name:               "case09_pSmaller_remainingLess_innerBufferEmpty",
+			bufferLen:          bufferLenSmaller,
+			remainingData:      remainingLess,
+			isInnerBufferEmpty: true,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			readLen, preConsume, totalLen := calculateParams(tc.bufferLen, tc.remainingData, tc.isInnerBufferEmpty, chunkSize)
+
+			// Arrange
+			data := sequentialBytes(totalLen)
+			checksumBytes := checksumBytesForData(data, chunkSize)
+			reader, err := csumio.NewReader(bytes.NewReader(data), bytes.NewReader(checksumBytes), chunkSize, true)
+			require.NoError(t, err)
+
+			if preConsume > 0 {
+				consumed := make([]byte, preConsume)
+				_, err := io.ReadFull(reader, consumed)
+				require.NoError(t, err)
+			}
+
+			remaining := totalLen - preConsume
+			require.Greater(t, remaining, 0, "test data must ensure bytes remain to read")
+
+			// Act
+			buf := make([]byte, readLen)
+			n, err := reader.Read(buf)
+			require.NoError(t, err)
+
+			// Assert
+			chunkStart := (preConsume / chunkSize) * chunkSize
+			chunkEnd := chunkStart + chunkSize
+			if chunkEnd > totalLen {
+				chunkEnd = totalLen
+			}
```

Comment:
Can we write this as follows?

```suggestion
			chunkEnd := min(chunkStart + chunkSize, totalLen)
```